File: cpp/001.TwoSum/TwoSum.cpp
// O(n)
// Runtime Beats 100.00%

#include <cassert>
#include <iostream>
#include <unordered_map>
#include <vector>

class Solution {
public:
    std::vector<int> twoSum(std::vector<int>& nums, int target) {
        std::unordered_map<int, int> map;
        int index = 0;
        for (const auto& elem : nums) {
            if (map.find(target - elem) != map.end()) {
                return {map[target - elem], index};
            } else {
                map[elem] = index;
            }
            index++;
        }
        return {};
    }
};

void test1() {
    std::vector<int> nums{2, 7, 11, 15};
    auto target = 9;
    std::vector<int> result{0, 1};
    Solution s;
    assert(s.twoSum(nums, target) == result);
}

void test2() {
    std::vector<int> nums{3, 2, 4};
    auto target = 6;
    std::vector<int> result{1, 2};
    Solution s;
    assert(s.twoSum(nums, target) == result);
}

void test3() {
    std::vector<int> nums{3, 3};
    auto target = 6;
    std::vector<int> result{0, 1};
    Solution s;
    assert(s.twoSum(nums, target) == result);
}

int main() {
    test1();
    test2();
    test3();
    return 0;
}

File: python3/001.TwoSum/TwoSum.py
# O(n)
# Runtime Beats 100.00%
from typing import List


class Solution:

    def twoSum(self, nums: List[int], target: int) -> List[int]:
        dict = {}
        for i, num in enumerate(nums):
            if target - num in dict:
                return [dict[target - num], i]
            else:
                dict[num] = i
        return []


def test1():
    s = Solution()
    nums = [2, 7, 11, 15]
    target = 9
    result = [0, 1]
    assert result == s.twoSum(nums, target)


def test2():
    s = Solution()
    nums = [3, 2, 4]
    target = 6
    result = [1, 2]
    assert result == s.twoSum(nums, target)


def test3():
    s = Solution()
    nums = [3, 3]
    target = 6
    result = [0, 1]
    assert result == s.twoSum(nums, target)


test1()
test2()
test3()
